tutorials/napps/switch_l3_part1.rst

```diff
@@ -0,0 +1,426 @@
+
+.. code-block:: bash
+
+  $ kytos napps create
+  --------------------------------------------------------------
+  Welcome to the bootstrap process of your NApp.
+  --------------------------------------------------------------
+  In order to answer both the username and the napp name,
+  You must follow this naming rules:
+   - name starts with a letter
+   - name contains only letters, numbers or underscores
+   - at least three characters
+  --------------------------------------------------------------
+
+  Please, insert your NApps Server username: tutorial
+  Please, insert your NApp name: of_l3ls
+  Please, insert a brief description for your NApp [optional]: This NApp does packet switching using L3 information
+
+  Congratulations! Your NApp have been bootstrapped!
+  Now you can go to the directory tutorial/of_l3ls and begin to code your NApp.
+  Have fun!
+
+Open the ``main.py`` file in your preferred editor to start coding your NApp.
+
+
+Create the switching table
+==========================
+
+First we will create an L3 switching table for each switch that connects to the
+controller. The table is implemented as a Python dictionary. The method needs
+a decorator in order to listen to the Kytos's *new switch* event.
+
+.. code-block:: python
+
+    @listen_to('kytos/core.switches.new')
+    def create_switching_table(self, event):
+        switch = event.content['switch']
+        switch.l3_table = {}
+
+.. ATTENTION:: Some classes and methods used in the code snippets need to be imported
+    in your main.py file. Please check the full content of the file and import these
+    elements as needed.
+
+
+Update switching table with incoming packets
+============================================
+
+In this step, you will create a method to handle PacketIn events. A PacketIn is
+generated everytime the switchs sends a packet to the controller for analysis.
+
+The method has a decorator to listen to *packet in* events. It unpacks the
+``packet_in.data`` into an Ethernet instance. If the frame data contains an IP
+packet, we unpack it into an IPv4 instance.
+
+Once you have the IPv4 header information, you can use it to update the
+switching table, associating it with the *in_port* - the physical switch
+interface where the packet came from.
+
+You can also add a log message to know when the controller receives a packet.
+
+.. code-block:: python
+
+    @listen_to('kytos/of_core.v0x01.messages.in.ofpt_packet_in')
+    def handle_packet_in(self, event):
+        packet_in = event.content['message']
+
+        ethernet = Ethernet()
+        ethernet.unpack(packet_in.data.value)
+
+        if ethernet.ether_type.value == 0x800:
+            ipv4 = IPv4()
+            ipv4.unpack(ethernet.data.value)
+
+            in_port = packet_in.in_port.value
+            switch = event.source.switch
+            switch.l3_table[ipv4.source] = in_port
+            log.info(f'Packet received from {ipv4.source} to {ipv4.destination}.')
```

Comment:
for python logging system it is recommended to pass the formatting string and the 'arguments' to be applied on the formatting string. This behavior enables python to build the string message to be outputed at the 'right time'.
`log.info("Packet received from %s to %s", ipv4.source, ipv4.destination)`